File: parameter_object_and_result_object_pattern/python/app1.py
# Define the UserParams class to encapsulate parameters
class UserParams:
    def __init__(self, name: str, age: int, address: str, email: str, phone: str):
        self.name = name
        self.age = age
        self.address = address
        self.email = email
        self.phone = phone


# Method to create a new user with UserParams object as a parameter
def create_user(user_params: UserParams):
    print(
        f"Creating user with name: {user_params.name}, age: {user_params.age}, address: {user_params.address}, email: {user_params.email}, phone: {user_params.phone}"
    )


# Method to update a user with UserParams object as a parameter
def update_user(user_id: int, user_params: UserParams):
    print(
        f"Updating user with ID: {user_id} with name: {user_params.name}, age: {user_params.age}, address: {user_params.address}, email: {user_params.email}, phone: {user_params.phone}"
    )


# Method to display user information
def display_user(user_params: UserParams):
    print(
        f"User Info - Name: {user_params.name}, Age: {user_params.age}, Address: {user_params.address}, Email: {user_params.email}, Phone: {user_params.phone}"
    )


if __name__ == "__main__":
    # Creating a UserParams object and passing it to the methods
    user_params = UserParams(
        name="John",
        age=30,
        address="123 Main St",
        email="john@example.com",
        phone="555-1234",
    )

    create_user(user_params)
    update_user(user_id=1, user_params=user_params)
    display_user(user_params)

File: parameter_object_and_result_object_pattern/typescript/app1.ts
// Define the UserParams class to encapsulate parameters
class UserParams {
    name: string;
    age: number;
    address: string;
    email: string;
    phone: string;

    constructor(name: string, age: number, address: string, email: string, phone: string) {
        this.name = name;
        this.age = age;
        this.address = address;
        this.email = email;
        this.phone = phone;
    }
}

// Method to create a new user with UserParams object as a parameter
function createUser(userParams: UserParams): void {
    console.log(
        `Creating user with name: ${userParams.name}, age: ${userParams.age}, address: ${userParams.address}, email: ${userParams.email}, phone: ${userParams.phone}`
    );
}

// Method to update a user with UserParams object as a parameter
function updateUser(userId: number, userParams: UserParams): void {
    console.log(
        `Updating user with ID: ${userId} with name: ${userParams.name}, age: ${userParams.age}, address: ${userParams.address}, email: ${userParams.email}, phone: ${userParams.phone}`
    );
}

// Method to display user information
function displayUser(userParams: UserParams): void {
    console.log(
        `User Info - Name: ${userParams.name}, Age: ${userParams.age}, Address: ${userParams.address}, Email: ${userParams.email}, Phone: ${userParams.phone}`
    );
}

// Main execution
(() => {
    // Creating a UserParams object and passing it to the methods
    const userParams = new UserParams(
        "John",
        30,
        "123 Main St",
        "john@example.com",
        "555-1234"
    );

    createUser(userParams);
    updateUser(1, userParams);
    displayUser(userParams);
})();
